src/nix/weather.cc

```diff
@@ -0,0 +1,191 @@
+                        attrPath2.push_back(attr);
+                        visit(*cursor2, attrPath2, false);
+                    }
+                };
+
+                if (cursor.isDerivation()) {
+                    auto drvPath = cursor.forceDerivation();
+                    drvPaths.insert(drvPath);
+                }
+
+                else if (
+                    attrPath.size() == 0
+                    || (attrPath[0] == "legacyPackages" && attrPath.size() <= 2)
+                    || (attrPath[0] == "packages" && attrPath.size() <= 2))
+                    recurse();
+
+                else if (initialRecurse)
+                    recurse();
+
+                else if (attrPath[0] == "legacyPackages" && attrPath.size() > 2) {
+                    auto attr = cursor.maybeGetAttr(state->sRecurseForDerivations);
+                    if (attr && attr->getBool())
+                        recurse();
+                }
+
+            } catch (EvalError & e) {
+                if (!(attrPath.size() > 0 && attrPath[0] == "legacyPackages"))
+                    throw;
+            }
+        };
+
+        for (auto & installable : installables) {
+            if (auto installable2 = std::dynamic_pointer_cast<InstallableFlake>(installable)) {
+                for (auto & [cursor, prefix] : installable2->getCursors(*state)) {
+                    visit(*cursor, parseAttrPath(*state, prefix), true);
+                }
+            } else {
+                auto drvPaths_ = toDerivations(store, {installable}, true);
+                drvPaths.insert(drvPaths_.begin(), drvPaths_.end());
+            }
+        }
+
+        if (drvPaths.size() == 0)
+            throw Error("no derivations found!");
+
+        StorePathSet closure;
+        if (!noClosure) {
+            Activity act(*logger, lvlInfo, actUnknown, "computing closure");
+            store->computeFSClosure(std::move(drvPaths), closure); // this could be huge ; should we chunk it up?
+        } else
+            closure = std::move(drvPaths);
+
+        StorePathSet outPaths;
+        for (auto & path : closure) {
+            Activity act(*logger, lvlInfo, actUnknown, fmt("querying '%s' output paths", store->printStorePath(path)));
+            if (isDerivation(store->printStorePath(path))) {
+                auto outputs = store->queryDerivationOutputs(path);
+                outPaths.insert(outputs.begin(), outputs.end());
+            }
+            // TODO: what to do with non-derivations?
+        }
+
+        StorePathSet totalValidPaths;
+        ssize_t totalPathsFound = 0;
+        uint64_t totalNarSize = 0;
+        std::optional<uint64_t> totalDownloadSize = 0;
+
+        for (auto & sub : subs) {
+            StorePathSet validPaths = sub->queryValidPaths(outPaths);
+
+            ssize_t pathsFound = 0;
+            uint64_t narSize = 0;
+            std::optional<uint64_t> downloadSize = 0;
+
+            for (auto & path : validPaths) {
```

Comment:
This looks very sequential so it will be latency-bound. Maybe you can use `queryMissing` instead? It should be a lot faster. Or use the async version of `queryPathInfo`.

Reply:
queryValidPaths should populate the cache up front. These queryPathInfo should be just cache lookups. I supposed some caches may not have cached backing them up? In that case, it would be very slow.